### ISSUE TITLE
[Application config] Usage and performance optimization

### ISSUE THREAD

**Issue by SuZhou-Joe** [opensearch-project/OpenSearch-Dashboards#6300]
**Is your feature request related to a problem? Please describe.**

## required parameters for  `getConfigurationClient` function
The [getConfigurationClient](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/application_config/server/plugin.ts#L53) method requires a scopedClusterClient to return the applicationConfigClient, but applicationConfig is supposed to be a generic config interface, which should not be related to opensearch.

## Performance concern on application config
The applicationClient.get is a async method which will fetch the doc inside `.opensearch-config` index. workspace or other plugins will leverage this service to get the `dynamic application settings`. In order to get latest config, plugins will call the function every time when there is a incoming request to handle, which may introduce a huge workload to the target storage.

**Describe the solution you'd like**

### required parameters for  `getConfigurationClient` function
change the required param to request: OpensearchDashboardRequest and construct the scopedClusterClient within OpenSearchConfigurationClient so that the implementation of the ConfigurationClient will be totally transparent to other plugins.

### Performance concern on application config
Add cache within application config. As for normal user case, user can only mutate the config through `application.updateEntityConfig` method, we can add a cache in memory to reduce the unnecessary call to the real storage service.

**Describe alternatives you've considered**

**Additional context**

#5855

**Comment by SuZhou-Joe** [comment]
@tianleh Would you like to take a look as you were the author of the related PR? cc @seraphjiang

**Comment by tianleh** [comment]
> @tianleh Would you like to take a look as you were the author of the related PR? cc @seraphjiang

checking

**Comment by kavilla** [comment]
+1 @SuZhou-Joe,

@tianleh. I think this should be a prioritization maybe even worth a patch for 2.13. If anyone is using this already.

As @SuZhou-Joe said I think we can utilize the storage service (which basically saves it it to local storage). If local storage does not contain the expect key then do the call dynamic app config. On app load we can do the update regardless and then subsequent requests will pull from local storage.

**Comment by SuZhou-Joe** [comment]
@kavilla Yes, for browser side we can do that but as the application config is a service for server side, I suppose we need to figure out a way to cache in server side as well.

**Comment by tianleh** [comment]
Created a RFC for the performance concern https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6341

Please review the proposal in the RFC

**Comment by seraphjiang** [comment]
> ## Issue 1: required parameters for `getConfigurationClient` function

how about this? could we have quick fix ?

**Comment by tianleh** [comment]
> how about this? could we have quick fix ?

It was implemented in such way to leverage the scoped client for permission check. I will run some experiments of the suggestion to `construct the scopedClusterClient within OpenSearchConfigurationClient`.

**Comment by tianleh** [comment]
Address both comments in this PR https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6364/files
